src/lib.rs

```diff
@@ -356,13 +360,26 @@ impl Discoverer {
         self
     }
 
+    /// Set which network interfaces to use for multicast joins.
+    ///
+    /// By default, multicast joins are performed only on the default interface.
+    /// Use `InterfaceSelection::All` to join on all suitable interfaces, or 
+    /// `InterfaceSelection::Specific(vec)` to join on specific interfaces by index.
+    ///
+    /// This can improve discovery in multi-homed environments where peers may
+    /// be on different network segments.
+    pub fn with_multicast_interfaces(mut self, selection: InterfaceSelection) -> Self {
+        self.interface_selection = selection;
+        self
+    }
+
     /// Start the discovery service.
     ///
     /// This will spawn asynchronous tasks and return a guard which will stop the discovery when dropped.
     /// Changing the configuration is done by stopping the discovery and starting a new one.
     pub fn spawn(self, handle: &Handle) -> Result<DropGuard, SpawnError> {
         let _entered = handle.enter();
-        let sockets = Sockets::new(self.class)?;
```

Comment:
Maybe it would be better if you could pass in the sockets that swarm discovery is supposed to use from the outside, and then all this interface enumeration & selection can live outside.
This might result in a much smaller PR that's way easier to review & accept.